castroulloaaaron - Exercise #1 Forward Proxy Caching

castroulloaaaron opened the pull request:
# Description
- Build a new service inside the contributors' directory, using your GitHub username as part of the path:
    - backend/contributors/<github_username>/dnd
- Add a list and a get endpoints
    - Both endpoints should follow the handler pattern, i.e. use POST http calls, where event equals the payload: `{"resource": "monsters"}`
    - The list event will receive "monsters" in the payload, while the get event will receive the monster's index: `{"monster_index":<index>}` 
- Using Python requests and your local DB, create a proxy caching service for both endpoints, so any data existing on your local DB will be retrieved from it, while the alternative will be to retrieve the data from the upstream service and save it into the local DB (caching the data).
- Validate the input from the request and the output from the response using a Marshmallow schema.

# Test plan

## Prerequisites
### Endpoints
- http://localhost:4000
- http://localhost:4000/monsters

### Bodies
#### Valid
- `{"resource": "monsters"}`
- `{"monster_index": "<index>"}`
#### Not Valid
- `{"resources": "monsters"}`
- `{"monsters_index": "<index>"}`
- `{}`
- ` `

### Headers
- Content-Type: application/json

### Verbs
- POST

### Examples
```bash
curl -X POST -H "Content-Type: application/json" -d '{"resource": "monsters"}' http://lo
calhost:4000
```

```bash
curl -X POST -H "Content-Type: application/json" -d '{"monster_index": "acolyte"}' http://lo
calhost:4000/monsters
```

## Test plan
1. Test both endpoints without data on the database
    - Database must be filled.
    - The response time is higher than the test below.
    - Get a valid response with the 200 code.
2. Test both endpoints with data
    - No API call should be made.
    - The response time decreases vs the test above.
    - Get a valid response with the 200 code.
3. Test with invalid input
    - Error message should be displayed.
5. Test with no input 
    - Error message should be displayed.
6. Test invalid index for the `http://localhost:4000/monsters` endpoint
    -  Not Found code/message should be displayed.

# Demo

## DB Migration 
<img width="1707" height="710" alt="Screenshot 2025-08-23 at 17 57 16" src="https://github.com/user-attachments/assets/b69b103e-3ea1-4303-bdae-34ab7c1c0937" />

## Demo videos for the endpoints functionality
- [Part 1: Everything except the Not Found](https://www.loom.com/share/342adfd204c24a5488bc5189fd932752?sid=3930826e-70d3-4323-825b-d6406c7648bd)
- [Part 2: Not Found and Happy Paths](https://www.loom.com/share/0dece32af4bb4d42b12f964f713d6fcb?sid=01c77e16-6b96-45c7-b0e9-6d22071bd80b)

# Improvements

- Unit/Integration/Load/Stress/Security/etc tests (probably covered in the next weeks).
- Add a TTL column that can be checked after fetching the data from the DB. If the current time is higher than the TTL then fetch from the service and store/replace/update the registry.
- Keep in memory the list of monsters. If the index is not on the list, return Not Found without calling the DB and the API.
- Create a memory cache with the N most requested monsters or the latest N requested monsters, to avoid hitting the DB. Purge can be done based on the strategy and the TTL.
- Keep a list of invalid indexes or create a Bloom Filter to return faster and avoid unnecessary calls to the DB and API.
- Use the Cache-Control header to cache locally the response for x time (based on TTL calculations or with a business requirement time). 
- Endpoint documentation (such as OpenAPI and the Swagger flavor).

# Bonus

Debugging files for VS Code. Once all the files are created, you need to use the `docker-compose.debug.yml` instead, like this: `docker-compose -f docker-compose.debug.yml up --build`. To attach your debugger, follow the instructions in the [VS Code Python debugging guide](https://code.visualstudio.com/docs/python/debugging) using the `Python Debugger: Debug using launch.json > Python Debugger: Remote Attach` option in the main.py file.

<img width="299" height="170" alt="image" src="https://github.com/user-attachments/assets/fcbbf221-7849-44b8-b48a-c119fa5d3cce" />

<img width="614" height="138" alt="image" src="https://github.com/user-attachments/assets/05e4ec4c-fc4e-4cb4-91ec-145f592fb800" />


**backend/run_local_debug.sh**
```bash
#!/bin/bash
set -e

DB_SCRIPT=${1:-populate_database}

if [ "$DB_SCRIPT" = "reset_db" ]; then
  python /app/framework/scripts/reset_db.py
  exit 0
else
  python /app/framework/scripts/populate_database.py
fi

pip install debugpy -t /tmp \
    && python /tmp/debugpy --wait-for-client --listen 0.0.0.0:5678 \
    -m flask run --host=0.0.0.0 --port=4000 --reload

```

**docker-compose.debug.yml**
```yaml
services:
  flask_app:
    build: 
      context: ./backend
      args:
        APPLICATION_FOLDER: ${APPLICATION_FOLDER}
    env_file:
      - ./.env
    ports:
      - "4000:4000"
      - "5678:5678"
    command: bash /app/run_local_debug.sh
    restart: unless-stopped
    volumes:
      - ./backend:/app
    depends_on:
      - flask_db        
  flask_db:
    image: postgres:12
    ports:
      - "5432:5432"
    environment:
      POSTGRES_HOST: flask_db
      POSTGRES_PORT: 5432
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: postgres
      POSTGRES_DB: postgres      
    volumes:
      - pgdata:/var/lib/postgresql/data
    restart: unless-stopped
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U postgres"]
      interval: 30s
      timeout: 10s
      retries: 5

volumes:
  pgdata: {}
```

**.vscode/launch.json**
```json
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Debugger: Remote Attach",
            "type": "debugpy",
            "request": "attach",
            "connect": {
                "host": "localhost",
                "port": 5678
            },
            "pathMappings": [
                {
                    "localRoot": "${workspaceFolder}/backend",
                    "remoteRoot": "/app"
                }
            ],
            "justMyCode": true
        }
    ]
}
```

castroulloaaaron commented on backend/contributors/castroulloaaaron/dnd/service.py:
```diff
@@ -0,0 +1,92 @@
+        except:
+            self._logger.warning(f'Error fetching the data for index {index} from the database')
+
+        data = self._api.get(index)
```
Sorry I'm missing one scenario here, if the monster requested is not found on the API, please let  me add this scenario before reviewing this.

Thank you

castroulloaaaron replied:
Done, ready to be reviewed.

diego-rivera-capdev reviewed (approved):
Simple approach, good responses, good documentation.

Crisarias reviewed (approved):
It looks good and align with the requirements. Just a note:

It is important for caching tables to have timestamp audit columns, you can use built-in functions for the update_at and create_at tables. These fields can be used later for this or another service to updated the cache.

castroulloaaaron commented:
Hi @Crisarias, I agreed with you if those tables represent entities on a non-cache real-world scenario. But since this is just a cache (this can be easily replaced with Redis, Memcache, in memory dictionary, you name it), which we do not care about (because of the nature of the cache), I don't want to add any kind of overengineering on the solution (KISS), reflected on the schemas (id + JSON object).

If the requirements change, then my approach will be to create a base class with those common fields (created, updated, deleted) and extend any entity from it.
